[InstCombine] Push freeze through non-recurrence PHIs

c-rhodes opened the pull request:
#157112 adds a test '@fold_phi_noundef_start_value' where we can't currently remove the freeze, even though it's possible. The two places in instcombine that deal with freezes are
'pushFreezeToPreventPoisonFromPropagating' and
'foldFreezeIntoRecurrence'.  The former doesn't support PHIs at all and the latter is restricted to recurrences. It doesn't consider this test a recurrence as the BB of the frozen PHI node '%loop.latch' does not dominate either of the BBs ('%loop', '%if.else') of the incoming values.

Therefore, I've updated 'pushFreezeToPreventPoisonFromPropagating' to support pushing the freeze to the incoming PHI values, as long as the BB of the frozen PHI *does not* dominate the BB of the incoming value(s). This fixes the test case added in #157112 and the other test changes look sensible, although perhaps I'm missing some edge cases (?).

The 'match(U.get(), m_Undef())' check for undef PHI incoming value looks necessary to catch the case covered by '@two_undef' test in freeze-phi.ll. Without this check the undef becomes zero which doesnt seem right.

c-rhodes committed:
[InstCombine] Push freeze through non-recurrence PHIs

PR #157112 adds a test '@fold_phi_noundef_start_value' where we can't
currently remove the freeze, even though it's possible. The two places
in instcombine that deal with freezes are
'pushFreezeToPreventPoisonFromPropagating' and
'foldFreezeIntoRecurrence'.  The former doesn't support PHIs at all and
the latter is restricted to recurrences. It doesn't consider this test a
recurrence as the BB of the frozen PHI node '%loop.latch' does not
dominate either of the BBs ('%loop', '%if.else') of the incoming values.

Therefore, I've updated 'pushFreezeToPreventPoisonFromPropagating' to
support pushing the freeze to the incoming PHI values, as long as the BB
of the frozen PHI *does not* dominate the BB of the incoming value(s).
This fixes the test case added in #157112 and the other test changes
look sensible, although perhaps I'm missing some edge cases (?).

The 'match(U.get(), m_Undef())' check for undef PHI incoming value looks
necessary to catch the case covered by '@two_undef' test in
freeze-phi.ll. Without this check the undef becomes zero which doesnt
seem right.

c-rhodes committed:
address comments

- remove unused %cond.2 from tests
- bail on backedge from BB of incoming value of phi to BB of freeze
- bail if incoming value of phi is from invoke and add a test
- bail if phi has multiple values from same predecessor and test

c-rhodes commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -5042,10 +5042,18 @@ InstCombinerImpl::pushFreezeToPreventPoisonFromPropagating(FreezeInst &OrigFI) {
 
+    if (auto *PN = dyn_cast<PHINode>(V)) {
+      if (llvm::any_of(PN->incoming_values(), [this, &PN](Use &U) {
+            return DT.dominates(PN->getParent(), PN->getIncomingBlock(U)) ||
```
added a check for `isBackEdge()`. Although unlike the other edge cases you pointed out I couldn't write a test case that triggered an issue for this one.

c-rhodes committed:
replace isa<InvokeInst> with generic check for terminator

c-rhodes commented:
ping

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -5042,10 +5042,29 @@ InstCombinerImpl::pushFreezeToPreventPoisonFromPropagating(FreezeInst &OrigFI) {
+            return false;
+        }
+
+        if (DT.dominates(BB, InBB) || isBackEdge(InBB, BB) ||
```
Is the dominates check here still needed?

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -5042,10 +5042,29 @@ InstCombinerImpl::pushFreezeToPreventPoisonFromPropagating(FreezeInst &OrigFI) {
+        }
+
+        if (DT.dominates(BB, InBB) || isBackEdge(InBB, BB) ||
+            VisitedBBs.contains(InBB) || match(U.get(), m_Undef()))
```
Use VisitedBBs.insert(InBB).second to avoid double HT lookup.

Also this deserves a comment. Ideally we would not have this limitation and make sure multi-edges use the same value.

c-rhodes committed:
address comments

c-rhodes committed:
address comments

c-rhodes committed:
Revert "address comments"

This reverts commit 790e7ede98c32b4a8c2f0fa69a2817edfc63e475.

c-rhodes commented:
Whilst doing some further testing I found 527.cam4_r in SPEC2017 is hanging with this patch. I see:
```
OrigUse->get() = <2 x i1> poison
OrigFI:   %.fr91682 = freeze <2 x i1> poison
OrigUse->get() = <2 x i1> poison
OrigFI:   %188 = freeze <2 x i1> %broadcast.splat2369
OrigUse->get() =   %broadcast.splat2369 = shufflevector <2 x i1> %broadcast.splatinsert2368, <2 x i1> poison, <2 x i32> zeroinitializer
```
is repeated when dumping OrigFI/OrigUse in `pushFreezeToPreventPoisonFromPropagating`, so I added check:
```diff --git a/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp b/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
index 6004feb1d885..be983ed63daf 100644
--- a/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
+++ b/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
@@ -5026,6 +5026,9 @@ InstCombinerImpl::pushFreezeToPreventPoisonFromPropagating(FreezeInst &OrigFI) {
       if (U == OrigUse)
         return nullptr;

+      if (match(V, m_Poison()))
+        return nullptr;
+
       auto *UserI = cast<Instruction>(U->getUser());
       if (auto *PN = dyn_cast<PHINode>(UserI))
         Builder.SetInsertPoint(PN->getIncomingBlock(*U)->getTerminator());
```

which fixes it, but still trying to write a test case. I did isolate the function in cam4 and attempt to reduce it will llvm-reduce, but just get testcases that take exactly N seconds to compile.

I also collected data on the # of freezes in SPEC2017 (Ofast, LTO) after this patch and there's a ~13% increase. #154336 also caused an increase of ~27%. Although as previously mentioned https://github.com/llvm/llvm-project/pull/154336#issuecomment-3201408296, I'm not convinced this is a particularly useful metric, but just wanted to share this.

c-rhodes committed:
fix compile-time issue in spec2017 527.cam4_r

c-rhodes commented:
I've rebased and added a fix for the compile-time issue in spec2017 527.cam4_r, although I'm unable to write a test for it. This IR hits the fix I've added: https://godbolt.org/z/KrTcWroqr, initially I bailed out here but it caused a regression, so I changed it to not freeze the poison value instead.

nikic commented:
I don't understand the fix you have applied. Doesn't this mean that we will now fail to freeze direct poison operands?

Can you provide some more context on what causes the compile-time issue?

c-rhodes commented:
> I don't understand the fix you have applied. Doesn't this mean that we will now fail to freeze direct poison operands?
>
> Can you provide some more context on what causes the compile-time issue?

Sorry I admit it's not ideal, but I'm at a bit of a loss as to what to do about it. I'm seeing lots of freezes being created on direct poison operands, e.g. `OrigFI:   %.fr91682 = freeze <2 x i1> poison` (notice the number in the var name). I thought perhaps freeze is being pushed around the loop, I've tried to reduce it and found it reduces to a testcase that takes as many seconds to compile as the timeout. The best I've managed to get is two tests, one that takes 5 seconds to compile and one 10 seconds:
1. https://gist.github.com/c-rhodes/37911b2b01fe450e23f9336a736193db
2. https://gist.github.com/c-rhodes/080fa33a2ffe1ed3b82beb5557b95b92

and the 10 second one only has 31 more lines than the 5 second one and looking at the diff quite a few more PHIs: https://gist.github.com/c-rhodes/9042f77580a24df6dfcd849b55d5de64

but I'm struggling to isolate the specific issue or create a small reproducer. The previous godbolt link i shared is from llvm-reduce, but the loops and control-flow is gone.

c-rhodes committed:
Revert "fix compile-time issue in spec2017 527.cam4_r"

This reverts commit 05d8a513fafc327bdf93935a5d23aafb6be8c7f1.

c-rhodes commented:
> I found reverting the change to `pushFreezeToPreventPoisonFromPropagating` in [[InstCombine] Allow freezing multiple operands #154336](https://github.com/llvm/llvm-project/pull/154336) to emit the freezes in a more natural order fixes the hang:
> 
> 
> ```
> diff --git a/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp b/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
> index 6004feb1d885..51739e818d59 100644
> --- a/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
> +++ b/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
> @@ -5041,7 +5047,7 @@ InstCombinerImpl::pushFreezeToPreventPoisonFromPropagating(FreezeInst &OrigFI) {
>        continue;
> 
>      // reverse() to emit freezes in a more natural order.
> -    for (Use &Op : reverse(I->operands())) {
> +    for (Use &Op : I->operands()) {
>        Value *OpV = Op.get();
>        if (isa<MetadataAsValue>(OpV) || isGuaranteedNotToBeUndefOrPoison(OpV))
>          continue;
> ```

@nikic any ideas here? Not sure if this is an acceptable fix.

c-rhodes commented:
> > @nikic any ideas here? Not sure if this is an acceptable fix.
> 
> I don't think so. This probably just fixes it by accident, by changing worklist order or something.
> 
> Something that may be worth doing (also independently of this PR) is teaching isGuaranteedNotToBePoison about the splat pattern. It includes two poison values, but only the poison-ness of the splatted value actually matters (plus some complications due to FMF probably).
> 
> Doing this might reduce the number of freezes we generate and the eliminate again. Though it doesn't really address whatever the underlying problem here is.

see https://github.com/llvm/llvm-project/pull/163570

It does reduce the number of freezes on poison values, for the 5s testcase comparing the debug output it reduces them by ~82%, but in terms of compile time it makes things slightly worse by increasing compile time to ~6.3s. I suspect because it's now less constrained and the analysis can go deeper. But as you say the there's a separate underlying issue here and I think detecting the splat pattern makes sense regardless.

c-rhodes committed:
Track visited PHI nodes

this fixes the hang in cam4.

c-rhodes committed:
Simplify track visited PHI nodes

this limits the change to pushFreezeToPreventPoisonFromPropagating and
also adds a debug statement when the cache is hit, along with a testcase
that triggers this.

c-rhodes committed:
Revert "Track visited PHI nodes"

This reverts commit 840c626bc249a5a904956c42f70be6992e4490a0.
